### PR TITLE
refactor: python js interface

### PR DESCRIPTION
**Pull request by codedealer**
Some changes to the `frontend` folder following the feedback in #343.

`css_and_js.py` has been cleaned up. There is only one helper method `call_SD` that will call any method in js SD object you specify. Additionally it will pass any arguments into the method via object so the names of python arguments are translated to js directly.

* Reliance on a shared property has been removed; every method now gets Gradio input argument as `params.x`.
* Methods are allowed to return nullish values now (although those are converted to empty arrays internally for compatibility with Gradio frontend code).

### CODE REVIEW

File: frontend/css_and_js.py
Comment (nderscore) [inline frontend/css_and_js.py:31]: I wonder if there's some JSON formatter we could use here instead to get character escaping automatically? This is probably fine for now though

**Comment by nderscore** [comment]
Thanks for taking a look into this and addressing my feedback. This is a much cleaner abstraction :clap: :tada: 

I left a small comment, but no big concerns here :+1:

**Comment by codedealer** [comment]
> I've opened a ticket to allow for global JS execution with Gradio folks, and would really love to wait and see what they come up with. [gradio-app/gradio#2141](https://github.com/gradio-app/gradio/issues/2141)

@altryne my guy, I literally got this merged a few days back gradio-app/gradio#2108 
As soon as 3.2 is released we shall have it.

**Comment by nderscore** [comment]
> In that case, why merfe a refactor now when the above will get released we'll have a much cleaner interface?

This change in https://github.com/gradio-app/gradio/pull/2108 only impacts how this library code gets initialized - not how method calls work. The changes in this PR should still be relevant.

**Comment by codedealer** [comment]
Well the interface will stay mostly the same and this refactor was done by me mainly in preparation for the upcoming merge (I started it after my PR in gradio was accepted).

What it will allow us to do is to remove a hack used to instantiate `SD` global object cleanly and remove that travesty:
```
load_detector = gr.Number(value=0, label="Load Detector", visible=False)
load_detector.change(None, None, None, _js=js(opt))
demo.load(lambda x: 42, inputs=load_detector, outputs=load_detector)
```

I understand that it's been very chaotic recently and I don't know how closely you've been paying attention to the dev repo's codebase but we actually write js code in actual js files now. So it's not have been for nothing

**Comment by altryne** [comment]
> write js code in actual js files now
That's an amazing upgrade on it's own, let's do it. the JS parts in python hurt my soul (and my vscode) 

> I understand that it's been very chaotic recently

Yeah, just a few short days, and tons of people asking for help or collaboration or access etc, good trouble.

**Comment by altryne** [comment]
@codedealer looks like this it? 
https://github.com/hlky/stable-diffusion-webui/pull/297 
I should set up codeowners so I get tag automatically for everything /frontend related 🤔

**Comment by codedealer** [comment]
Yes, this PR builds on top of that one.

**Comment by altryne** [comment]
Catching up

File: frontend/css_and_js.py
Comment (altryne) [inline frontend/css_and_js.py:28]: Should this be renamed callJS ? 
callSD is ambiguous (SD refers to a lot of things in this field)

Comment (altryne) [reply frontend/css_and_js.py:28]: Could be a great place for 1 usage example as well

File: frontend/frontend.py
Comment (altryne) [inline frontend/frontend.py:54]: Can we do something about this double string requirement in fromId?

**Review by altryne — CHANGES_REQUESTED** [review]

File: frontend/css_and_js.py
Comment (codedealer) [reply frontend/css_and_js.py:28]: SD is the global js object. The logic behind this is that the whole line in code will look like so:
`_js=call_SD('foo')` will translate in js into `SD.foo()`
I'd argue `_js=call_JS` is ambiguous because it's not telegraphed what where will that call go.

File: frontend/frontend.py
Comment (nderscore) [reply frontend/frontend.py:54]: Related: https://github.com/hlky/stable-diffusion-webui/pull/381#discussion_r959814261

This could be handled here ^ if we accept a dictionary for arguments and convert it into a JSON string.

Comment (codedealer) [reply frontend/frontend.py:54]: We can assume that every argument is a string but I really wish to retain the ability to pass other types.
See line 70 where I pass an object instead

File: frontend/css_and_js.py
Comment (altryne) [reply frontend/css_and_js.py:28]: It's no ambiguous in the python context, as _js is a python attribute, and call_JS is a python method, and what it does, it calling a JS function. 
Naming the global object SD is confusing to me, please address

Comment (codedealer) [reply frontend/css_and_js.py:28]: I can rename python's `call_SD` to `call_JS` no problem. What should be a global object named be then if not SD? GradioHelper?

File: frontend/frontend.py
Comment (nderscore) [reply frontend/frontend.py:54]: @codedealer 

I would do it like this:

```py
import json

some_options = {
  "num": 123,
  "bool": True,
  "str": "this is a string"
}

json_str = json.dumps(some_options)

function_name = "some_function"

function_call_str = f"async (x) => {{ return await SD.{function_name}({{ ...{json_str}, x }}) ?? [] }}"
```

File: frontend/css_and_js.py
Comment (hlky) [reply frontend/css_and_js.py:28]: I agree with @codedealer it makes more sense to be named `call_SD`

File: frontend/frontend.py
Comment (hlky) [reply frontend/frontend.py:54]: Types are useful.

File: frontend/css_and_js.py
Comment (nderscore) [reply frontend/css_and_js.py:31]: Copying this from the other comment chain:

I would do recommend doing it like this:

```py
import json

some_options = {
  "num": 123,
  "bool": True,
  "str": "this is a string"
}

json_str = json.dumps(some_options)

function_name = "some_function"

function_call_str = f"async (x) => {{ return await SD.{function_name}({{ ...{json_str}, x }}) ?? []; }}"
```

which results in:

```js
async (x) => { return await SD.some_function({ ...{"num": 123, "bool": true, "str": "this is a string"}, x }) ?? [] }
```

which is essentially the same as this after the object spread is evaluated:

```js
async (x) => { return await SD.some_function({ num: 123, bool: true, str: "this is a string", x }) ?? [] }
```

And the original types are retained without needing to manually add quotes around strings or convert `True` to `true` etc.

File: frontend/frontend.py
Comment (codedealer) [reply frontend/frontend.py:54]: @nderscore I phrased this poorly, in fact I'm not sure how to describe it so I'll provide a real example:

 How do I get a reference to a js object in this way? dict passed in `json.dumps()` requires all arguments to be valid python values.
Currently I have this
`call_SD("textToClipboard", text="x")`
where x references _the js object that does not exist during python's execution_. With proposed way this `"x"` argument will be treated as a string by JSON and it will subsequently pass a string to your js method when you wanted an object reference.

Comment (nderscore) [reply frontend/frontend.py:54]: In my example, the `x` is still being proxy-passed through like before.

Shoudn't `SD.textToClipboard` read that value from `{ x }` instead?

```py
# text to clipboard only uses `x` from gradio, so it doesn't need any custom arguments
call_SD("textToClipboard")
```
```js
// no options should be an empty object:
async (x) => { return await SD.textToClipboard({ ...{}, x }) ?? []; }
// which is interpretted like this:
async (x) => { return await SD.textToClipboard({ x })
// and then the textToClipboard method should use x:
textToClipboard({ x }) {
  // copy x to clipboard
}
```

Comment (codedealer) [reply frontend/frontend.py:54]: Maybe doing things like `text="x"` is not such a great idea? 🤔

Comment (codedealer) [reply frontend/frontend.py:54]: @nderscore I tried to keep `textToClipboard()` argument agnostic in case we use it in the future from js itself. But yeah, technically speaking it doesn't need any arguments right now.

Comment (nderscore) [reply frontend/frontend.py:54]: You could have two functions: One that's a generic utility function that takes a string, and one that's a wrapper around that utility function which consumes `x` from gradio

Comment (codedealer) [reply frontend/frontend.py:54]: working on it as we speak!

**Review by altryne — APPROVED** [review]

**Comment by codedealer** [comment]
@altryne don't merge it yet :D

**Comment by codedealer** [comment]
Clipboard api, gave me grief. Should be good now

File: frontend/css_and_js.py
Comment (nderscore) [reply frontend/css_and_js.py]: This was resolved by [4da7cb6](https://github.com/hlky/stable-diffusion-webui/pull/381/commits/4da7cb643723c8fb7c5f9f9a06353b709534dd79)

**Comment by nderscore** [comment]
updates in [4da7cb6](https://github.com/hlky/stable-diffusion-webui/pull/381/commits/4da7cb643723c8fb7c5f9f9a06353b709534dd79) look great :clap: 

thanks again for updating :pray:

**Pull request merged**
